### DISCUSSION TITLE
Could not resolve dependency when upgrading from Next.js 14 to 15 RC

### DISCUSSION THREAD

**Discussion by tjhoo** [vercel/next.js#66259] [Help]
### Summary

Hi,

I'm trying to follow the steps [here](https://rc.nextjs.org/docs/app/building-your-application/upgrading/version-15) to upgrade Next.js 14 to 15 RC and encounter the following error
```
$ npm install next@rc react@rc react-dom@rc
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR!
npm ERR! While resolving: dummy@0.1.0
npm ERR! Found: react@19.0.0-rc-935180c7e0-20240524
npm ERR! node_modules/react
npm ERR!   react@"19.0.0-rc-935180c7e0-20240524" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peer react@"19.0.0-rc-f994737d14-20240522" from next@15.0.0-rc.0
npm ERR! node_modules/next
npm ERR!   next@"15.0.0-rc.0" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR!
```
is this an expected error and we should specify `--force` when installing the RC packages?
```
npm install --force next@rc react@rc react-dom@rc
```

**Comment by icyJoseph** [comment]
Hi,

The problem is that, https://github.com/vercel/next.js/blob/canary/packages/next/package.json#L108-L109, lists those RC specific versions, as peerDependencies.

Eventually, when React 19 does stabilize, this won't be a problem, but as release day approaches, and new react, react-dom RC versions are publishes, Next.js is bound to lag slightly behind.

**Reply by tjhoo** [reply]
Yes, and we have specify `--force` when upgrading to Next.js 15.

The command has to be fixed in [the upgrade guide](https://rc.nextjs.org/docs/app/building-your-application/upgrading/version-15)

**Reply by judygab** [reply]
Specifying `--force` didn't work for me either, I still got errors. Is there anything else you tried?

**Reply by stoufa88** [reply]
I tried upgrading a barebone next 14 app to 15 and it worked with `--force`. pnpm also worked for me while showing these warnings:

` WARN  Issues with peer dependencies found
.
└─┬ next 15.0.0-rc.0
  ├── ✕ unmet peer react@19.0.0-rc-f994737d14-20240522: found 19.0.0-rc.0
  └── ✕ unmet peer react-dom@19.0.0-rc-f994737d14-20240522: found 19.0.0-rc.0`

**Reply by admintvalley** [reply]
same warning ┬ next 15.0.0-rc.0
  ├── ✕ unmet peer react@19.0.0-rc-f994737d14-20240522: found 19.0.0-rc-6230622a1a-20240610
  └── ✕ unmet peer react-dom@19.0.0-rc-f994737d14-20240522: found 19.0.0-rc-6230622a1a-20240610

**Reply by AshishKalaria** [reply]
> Specifying `--force` didn't work for me either, I still got errors. Is there anything else you tried?

first delete node_modules folder and package-lock.json file and try that command

**Comment by gumpsh** [comment]
Has anyone managed to sort this yet? npm error code ERESOLVE
npm error ERESOLVE could not resolve
npm error
npm error While resolving: @heroicons/react@2.1.4
npm error Found: react@19.0.0-rc-6230622a1a-20240610
npm error node_modules/react
npm error   react@"19.0.0-rc-6230622a1a-20240610" from the root project
npm error   peer react@"19.0.0-rc-6230622a1a-20240610" from react-dom@19.0.0-rc-6230622a1a-20240610
npm error   node_modules/react-dom
npm error     react-dom@"19.0.0-rc-6230622a1a-20240610" from the root project
npm error   1 more (styled-jsx)
npm error
npm error Could not resolve dependency:
npm error peer react@">= 16" from @heroicons/react@2.1.4
npm error node_modules/@heroicons/react
npm error   @heroicons/react@"^2.1.4" from the root project
npm error
npm error Conflicting peer dependency: react@18.3.1
npm error node_modules/react
npm error   peer react@">= 16" from @heroicons/react@2.1.4
npm error   node_modules/@heroicons/react
npm error     @heroicons/react@"^2.1.4" from the root project
npm error
npm error Fix the upstream dependency conflict, or retry
npm error this command with --force or --legacy-peer-deps
npm error to accept an incorrect (and potentially broken) dependency resolution.
npm error
npm error
npm error For a full report see:

I have been trying all sorts of flags like --force, --legacy-peer-deps, -npm reinstall, --no-frozen-lockfile. I have also tried manually changing versions in the package.json, package-lock.json and yaml files. Nothing seems to work and my deployments are failing.

This is the dashboard project on the next.js getting started tutorial by the way.

I can launch the project from vscode and it all seems to run ok locally - but failing on the build I guess.

**Reply by o1Suleyman** [reply]
Still an issue in November. npm install --force works but there's a lot of warnings

**Reply by icyJoseph** [reply]
Just use the cli mod to get to next@15, but I'd rather wait until React 19 comes out officially 

- https://nextjs.org/blog/next-15#smooth-upgrades-with-nextcodemod-cli

**Reply by o1Suleyman** [reply]
> Just use the cli mod to get to next@15, but I'd rather wait until React 19 comes out officially
> 
>     * https://nextjs.org/blog/next-15#smooth-upgrades-with-nextcodemod-cli

I wasn't upgrading though, I got these errors starting a new project with the instructions laid out here: https://nextjs.org/learn/dashboard-app/getting-started (sorry if I'm misunderstanding, I'm new to the JS ecosystem)

**Reply by gumpsh** [reply]
You don't have to be upgrading for these issues to arise. Correct me if I'm wrong, but I believe its because next have set the project up to use react 19c, but when to run npm install, some other dependencies in the project are not compatible with 19c. This is why bringing the react version back to a stable release version resolves the issue.

**Reply by icyJoseph** [reply]
Yeah, something along those lines. It didn't end there for me though. Once I had it all running, I had some type issues coming from 19 type changes, which are not very nice to fix 😅

**Reply by dusanbrankov** [reply]
I had the same problem yesterday and was able to solve it by adding the following to package.json:

```json
  "overrides": {
    "some-dependency": {
      "react": "$react"
    }
  }
```

I had problems with `drizzle-orm` and `react-hook-form`, so my overrides directive looks like this:

```json
  "overrides": {
    "drizzle-orm": {
      "react": "$react"
    },
    "react-hook-form": {
      "react": "$react"
    }
  }
```

Reference: https://github.com/drizzle-team/drizzle-orm/issues/2251#issuecomment-2435153847

**Reply by o1Suleyman** [reply]
> You don't have to be upgrading for these issues to arise. Correct me if I'm wrong, but I believe its because next have set the project up to use react 19c, but when to run npm install, some other dependencies in the project are not compatible with 19c. This is why bringing the react version back to a stable release version resolves the issue.

quite crazy to me coming from Java but otherwise I'm really enjoying Next

**Reply by icyJoseph** [reply]
Yeah, the React 19 release hit a major bump some months ago, and it has taken a bit to figure out a solution, plus it's implementation. Next 15 was kind of hoping that 19 would be out many months ago.

**Reply by gumpsh** [reply]
Npm dependencies - fun times! I've been developing a react native app in work and took a quick udemy course to learn the basics first. Anyway, a few months later I came back to the course to check something out and it just blew up everywhere after I decided to update the expo version. I then had to go through my dependencies in package.json to fix it. Expo was very good at telling me which dependencies to fix though and even told me which versions to update to.

**Comment by laurenmacam** [comment]
using pnpm this worked for me:
delete pnpm-lock.yaml file
run
pnpm up next@rc react@rc react-dom@rc eslint-config-next@rc
pnpm i

**Reply by gumpsh** [reply]
Hi @laurenmacam - thanks for your suggestion. I did try it, but still getting issues unfortunately.

However, I finally managed to solve it by downgrading from next, react and react-dom rc versions to latest stable releases.

It was a real mission, but all test are passing on the Vercel deployment now, so I am happy!

**Comment by jastka4** [comment]
Adding this to my `package.json` helped with all the issues I had:
```json
  "overrides": {
    "next": {
      "react": "^19.0.0-rc-a960b92c-20240819",
      "react-dom": "^19.0.0-rc-a960b92c-20240819"
    }
  }
```

You might want to change the version to the latest React RC version.

**Reply by gumpsh** [reply]
Yeah, will definitely be upgrading to latest version as soon as everything catches up with it - which may be the case already for all I know - I've been doing other things for a while!

**Comment by bugplanter** [comment]
`npm i next@rc react@rc react-dom@rc eslint-config-next@rc --legacy-peer-deps`

**Reply by jdhiro** [reply]
There is a reason legacy-peer-deps are legacy. This is a bad practice.

You'd be better off creating an overrides section in your package.json to deal with specific version overrides.

**Comment by jdhiro** [comment]
Well I really want to upgrade, but I'm just taking this as a sign to wait until the next RC at least.

**Reply by RafaelZasas** [reply]
Dont do it. I'm in dependency hell atm. I just want to go home...

**Reply by gumpsh** [reply]
Is react 19 a stable release yet?

**Comment by JackGarrus** [comment]
Hello, this seems to be still an issue for me, as you can see:

<img width="608" alt="Screenshot 2025-04-19 alle 14 00 03" src="https://github.com/user-attachments/assets/4ce75af1-2c43-4751-9b41-f32cdba27387" />


Since React 19 is now stable for a few months, is there already a workaround/solution to this issue that I might have been missed? I **really** don't want to use `--legacy-peer-deps`...

**Reply by gumpsh** [reply]
I just had a quick Google for next mdx remote compatibility with react 19 and found this:

https://github.com/hashicorp/next-mdx-remote/blob/main/CHANGELOG.md

Seems to suggest upgrading to v5.0.0
